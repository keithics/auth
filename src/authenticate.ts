import jwt from 'jsonwebtoken';
import { catchAsync } from '@keithics/errors/lib/catch-async';
import { arkAssert, ArkErrorInvalidToken } from '@keithics/errors/lib/ark.assert';

interface AuthenticationInterface {
  token: string;
  level: string;
  roles: string[];
  exactRoles?: boolean;
  allUsers?: boolean;
}

type AuthenticationUserType = { jwt: string } & AuthenticationInterface;

interface AccessInterface {
  userRoles: string[];
  exactRoles: boolean;
  roles: string[];
}

/**
 * Authenticate JWT
 * @param token - JWT token
 * @param level - user level eg: user or superadmin
 * @param roles - needed roles, array of user roles, skipped checking if super admin or roles has ['*']
 * @param exactRoles - if true, all roles must be satisfied otherwise use false if only a subset of roles is found
 *
 * eg:
 * exactRoles = false
 * user roles = ['a','b'] and neededRoles are ['a','b','c'] == returns true
 * exactRoles = true
 * user roles = ['a','b'] and neededRoles are ['a','b','c'] == returns false, must be exact and lacks 'c' role
 * @param allUsers - no checking of roles needed
 */

export const authenticate = function ({ token, level, roles, exactRoles = true, allUsers = false }: AuthenticationInterface) {
  return catchAsync(async (req, res, next): Promise<void> => {
    const jwt = req.headers?.authorization?.replace('Bearer ', '') || null;
    const decoded = authenticateUser({ jwt, token, level, roles, exactRoles, allUsers });
    req.user = decoded.user;
    next();
  });
};

/***
 * Separated function for easier testing
 */
export const authenticateUser = ({ jwt: jwtUser, token, level, roles, exactRoles = true, allUsers = false }: AuthenticationUserType) => {
  arkAssert(jwtUser, ArkErrorInvalidToken, 'JWT Token is invalid');
  const decoded = jwt.verify(jwtUser, token);
  arkAssert(decoded?.user.level === level, ArkErrorInvalidToken, 'user token is invalid');
  const { level: userLevel, roles: userRoles } = decoded?.user;

  if (userLevel !== 'superadmin') {
    arkAssert(userRoles, ArkErrorInvalidToken, 'user role is empty');
    if (!userRoles.includes('*') && !allUsers) {
      const isDenied = isAccessDenied({ exactRoles, userRoles, roles });
      arkAssert(isDenied, ArkErrorInvalidToken, 'user role is not allowed');
    }
  }

  return decoded;
};

const isAccessDenied = ({ exactRoles, userRoles, roles }: AccessInterface): boolean => {
  let accessDenied;
  if (exactRoles === false) {
    accessDenied = roles.some((r) => userRoles.includes(r));
  } else {
    accessDenied = roles.every((r) => userRoles.includes(r));
  }

  return accessDenied;
};
