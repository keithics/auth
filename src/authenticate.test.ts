import jwt from 'jsonwebtoken';
import { authenticateUser } from './authenticate';
import { ArkErrorInvalidToken } from '@keithics/errors/lib/ark.assert';

const body = {
  user: {
    _id: '61a9143ff6d9d2000dc3bceb',
    email: 'keith.lumanog@unitedintegral.com',
    level: 'user',
    roles: ['a', 'b', 'z'],
  },
};

const token = 'token';
const jwtUser = jwt.sign(body, token);

body.user.roles = ['*'];
const jwtUserAllRoles = jwt.sign(body, token);

delete body.user.roles;
const jwtUserNoRoles = jwt.sign(body, token);

body.user.level = 'superadmin';
const jwtAdmin = jwt.sign(body, token);

describe('Authentication test', () => {
  test('Should be ðŸ˜ž if user has no roles', async () => {
    expect(() => {
      authenticateUser({ jwt: jwtUserNoRoles, token, level: 'user', roles: ['c'] });
    }).toThrowError(ArkErrorInvalidToken);
  });

  test('Should be ðŸ˜ž if user has no access, needs to have c role', async () => {
    expect(() => {
      authenticateUser({ jwt: jwtUserNoRoles, token, level: 'user', roles: ['c'] });
    }).toThrowError(ArkErrorInvalidToken);
  });

  test('Should be ðŸ‘Œ if user has SOME of the access, common role is b', async () => {
    expect(authenticateUser({ jwt: jwtUser, token, level: 'user', roles: ['c', 'b'], exactRoles: false })).toMatchSnapshot();
  });

  test('Should be ðŸ‘Œ if user has SOME of the access, common role is b and z', async () => {
    expect(
      authenticateUser({
        jwt: jwtUser,
        token,
        level: 'user',
        roles: ['c', 'b', 'z'],
        exactRoles: false,
      })
    ).toMatchSnapshot();
  });

  test('Should be ðŸ‘Œ if user has ALL a,b and z roles ', async () => {
    expect(authenticateUser({ jwt: jwtUser, token, level: 'user', roles: ['a', 'b', 'z'] })).toMatchSnapshot();
  });

  test('Should be ðŸ˜ž if user has SOME a and b roles but no c role ', async () => {
    expect(() => {
      authenticateUser({ jwt: jwtUser, token, level: 'user', roles: ['a', 'b', 'c'] });
    }).toThrowError(ArkErrorInvalidToken);
  });

  test('Should be ðŸ˜ž if user has ALL a,b and z roles but no c role ', async () => {
    expect(() => {
      authenticateUser({ jwt: jwtUser, token, level: 'user', roles: ['a', 'b', 'z', 'c'] });
    }).toThrowError(ArkErrorInvalidToken);
  });

  test('Should be ðŸ‘Œ if user is a superadmin, no roles needed', async () => {
    expect(authenticateUser({ jwt: jwtAdmin, token, level: 'superadmin', roles: [] })).toMatchSnapshot();
  });

  test("Should be ðŸ‘Œ if even if the user doesn't have the role because allUsers is true", async () => {
    expect(authenticateUser({ jwt: jwtUser, token, level: 'user', roles: ['x'], allUsers: true })).toMatchSnapshot();
  });

  test('Should be ðŸ‘Œ if even if the user role is equal to *', async () => {
    expect(authenticateUser({ jwt: jwtUserAllRoles, token, level: 'user', roles: ['x'] })).toMatchSnapshot();
  });

  test('Should be ðŸ‘Œ if even if the user role is equal to * and allUsers: true', async () => {
    expect(authenticateUser({ jwt: jwtUserAllRoles, token, level: 'user', roles: ['x'], allUsers: true })).toMatchSnapshot();
  });
});
