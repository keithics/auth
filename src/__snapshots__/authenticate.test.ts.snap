// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication test Should be ðŸ‘Œ if even if the user doesn't have the role because allUsers is true 1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "user",
    "roles": Array [
      "a",
      "b",
      "z",
    ],
  },
}
`;

exports[`Authentication test Should be ðŸ‘Œ if even if the user role is equal to * 1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "user",
    "roles": Array [
      "*",
    ],
  },
}
`;

exports[`Authentication test Should be ðŸ‘Œ if even if the user role is equal to * and allUsers: true 1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "user",
    "roles": Array [
      "*",
    ],
  },
}
`;

exports[`Authentication test Should be ðŸ‘Œ if user has ALL a,b and z roles  1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "user",
    "roles": Array [
      "a",
      "b",
      "z",
    ],
  },
}
`;

exports[`Authentication test Should be ðŸ‘Œ if user has SOME of the access, common role is b 1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "user",
    "roles": Array [
      "a",
      "b",
      "z",
    ],
  },
}
`;

exports[`Authentication test Should be ðŸ‘Œ if user has SOME of the access, common role is b and z 1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "user",
    "roles": Array [
      "a",
      "b",
      "z",
    ],
  },
}
`;

exports[`Authentication test Should be ðŸ‘Œ if user is a superadmin, no roles needed 1`] = `
Object {
  "iat": 1654112080,
  "user": Object {
    "_id": "61a9143ff6d9d2000dc3bceb",
    "email": "keith.lumanog@unitedintegral.com",
    "level": "superadmin",
  },
}
`;
